src/itemstackmetadata.cpp

```diff
@@ -1,34 +1,57 @@
 #include "itemstackmetadata.h"
 #include "util/serialize.h"
 #include "util/strfnd.h"
+#include <json/json.h>
 
-#define DESERIALIZE_START '\x01'
-#define DESERIALIZE_KV_DELIM '\x02'
-#define DESERIALIZE_PAIR_DELIM '\x03'
-#define DESERIALIZE_START_STR "\x01"
-#define DESERIALIZE_KV_DELIM_STR "\x02"
-#define DESERIALIZE_PAIR_DELIM_STR "\x03"
+#define SERIALIZATION_VERSION_IDENTIFIER '\x02'
 
 void ItemStackMetadata::serialize(std::ostream &os) const
 {
-	std::ostringstream os2;
-	os2 << DESERIALIZE_START;
+	Json::Value root(Json::objectValue);
 	for (StringMap::const_iterator
 			it = m_stringvars.begin();
 			it != m_stringvars.end(); ++it) {
-		os2 << it->first << DESERIALIZE_KV_DELIM
-		    << it->second << DESERIALIZE_PAIR_DELIM;
+		root[it->first] = it->second;
 	}
-	os << serializeJsonStringIfNeeded(os2.str());
+	os << SERIALIZATION_VERSION_IDENTIFIER + Json::FastWriter().write(root);
 }
 
+#define DESERIALIZE_START '\x01'
+#define DESERIALIZE_KV_DELIM '\x02'
+#define DESERIALIZE_PAIR_DELIM '\x03'
+#define DESERIALIZE_START_STR "\x01"
+#define DESERIALIZE_KV_DELIM_STR "\x02"
+#define DESERIALIZE_PAIR_DELIM_STR "\x03"
+
 void ItemStackMetadata::deSerialize(std::istream &is)
 {
-	std::string in = deSerializeJsonStringIfNeeded(is);
-
 	m_stringvars.clear();
 
-	if (!in.empty()) {
+	// Check for EOF
+	if (!is.good())
+		return;
+
+	// Decode Json metadata (ItemMeta 3, 2017-07-02)
+	if (is.peek() == SERIALIZATION_VERSION_IDENTIFIER) {
+		is.ignore(1);
+		std::string in(std::istreambuf_iterator<char>(is), {});
+
+		Json::Reader reader;
+		Json::Value attr_root;
+		if (reader.parse(in, attr_root) && attr_root.type() == Json::objectValue) {
+			const Json::Value::Members attr_list = attr_root.getMemberNames();
+			for (auto it : attr_list) {
```

Comment:
if i remember it's possible to do for (const auto &attr : attr_root) { attr.asString(); }